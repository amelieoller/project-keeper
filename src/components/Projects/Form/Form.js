import React, { Component } from 'react';
import styled from 'styled-components';
import Button from '../../Button';
import Input from './Input';

const StyledForm = styled.div`
  position: relative;
  max-width: 40em;
  margin: 5em auto;
  background: #fff;
  width: 100%;
  padding: 3em 5em;
  border-radius: 3px;

  &::before {
    content: '';
    position: absolute;
    top: 0;
    left: 0;
    right: 0;
    bottom: 0;
    box-shadow: 0 4px 10px 1px rgba(0, 0, 0, 0.14),
      0 3px 14px 2px rgba(0, 0, 0, 0.12), 0 5px 5px -3px rgba(0, 0, 0, 0.2);
    transform: scale(0.98);
    transition: transform 0.28s ease-in-out;
    z-index: -1;
  }
`;

class Form extends Component {
  state = {
    title: '',
    description: '',
    github: '',
    folder: '',
    image: '',
    server: ''
  };

  render() {
    const { handleSubmit } = this.props;

    return (
      <StyledForm>
        <form
          onSubmit={e => {
            e.preventDefault();
            const {
              title,
              description,
              folder,
              github,
              image,
              server
            } = e.target.elements;

            handleSubmit({
              title: title.value,
              description: description.value,
              folder: folder.value,
              github: github.value,
              image: image.value,
              server: server.value
            });
          }}
          data-testid="form"
        >
          <Input name="title" placeholder="Title" required />
          <Input name="description" placeholder="Description" required />
          <Input name="github" placeholder="GitHub URL" />
          <Input name="folder" placeholder="Folder Name" />
          <Input name="image" placeholder="Image" />
          <Input name="server" placeholder="Server Command" />
          <Button type="submit">Submit</Button>
        </form>
      </StyledForm>
    );
  }
}

export default Form;
