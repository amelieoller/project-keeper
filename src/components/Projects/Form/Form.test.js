import React from 'react';
import { render, fireEvent } from '@testing-library/react';
import Form from './Form';
const faker = require('faker');

test('calls onSubmit with the appropriate fields when submitted', () => {
  // Arrange
  const handleSubmit = jest.fn();
  const { getByTestId, getByLabelText, getByText } = render(
    <Form handleSubmit={handleSubmit} />
  );

  const fakeInput = {
    title: faker.lorem.word(),
    description: faker.lorem.paragraph(),
    github: faker.internet.url(),
    folder: faker.lorem.word(),
    image: faker.lorem.word(),
    server: faker.lorem.word()
  };

  const titleNode = getByLabelText(/title/i);
  const descriptionNode = getByLabelText(/description/i);
  const githubNode = getByLabelText(/github/i);
  const folderNode = getByLabelText(/folder/i);
  const imageNode = getByLabelText(/image/i);
  const serverNode = getByLabelText(/server/i);
  const submitButtonNode = getByText(/submit/i);
  const formNode = getByTestId('form');

  // Act
  titleNode.value = fakeInput.title;
  descriptionNode.value = fakeInput.description;
  githubNode.value = fakeInput.github;
  folderNode.value = fakeInput.folder;
  imageNode.value = fakeInput.image;
  serverNode.value = fakeInput.server;
  fireEvent.submit(formNode);

  // Assert
  expect(handleSubmit).toHaveBeenCalledTimes(1);
  expect(handleSubmit).toHaveBeenCalledWith(fakeInput);
  expect(submitButtonNode.type).toBe('submit');
});

test('snapshot', () => {
  const { container } = render(<Form />);
  expect(container).toMatchSnapshot();
});
