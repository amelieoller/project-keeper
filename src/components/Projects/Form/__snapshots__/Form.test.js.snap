// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshot 1`] = `
<div>
  <div
    class="sc-htpNat eCNowg"
  >
    <form
      data-testid="form"
    >
      <div
        class="sc-bwzfXH jqqosM"
      >
        <input
          id="title"
          name="title"
          required=""
          type="text"
        />
        <label
          for="title"
        >
          Title
        </label>
        <i
          class="bar"
        />
      </div>
      <div
        class="sc-bwzfXH jqqosM"
      >
        <input
          id="description"
          name="description"
          required=""
          type="text"
        />
        <label
          for="description"
        >
          Description
        </label>
        <i
          class="bar"
        />
      </div>
      <div
        class="sc-bwzfXH jqqosM"
      >
        <input
          id="github"
          name="github"
          required=""
          type="text"
        />
        <label
          for="github"
        >
          GitHub URL
        </label>
        <i
          class="bar"
        />
      </div>
      <div
        class="sc-bwzfXH jqqosM"
      >
        <input
          id="folder"
          name="folder"
          required=""
          type="text"
        />
        <label
          for="folder"
        >
          Folder Name
        </label>
        <i
          class="bar"
        />
      </div>
      <div
        class="sc-bwzfXH jqqosM"
      >
        <input
          id="image"
          name="image"
          required=""
          type="text"
        />
        <label
          for="image"
        >
          Image
        </label>
        <i
          class="bar"
        />
      </div>
      <div
        class="sc-bwzfXH jqqosM"
      >
        <input
          id="server"
          name="server"
          required=""
          type="text"
        />
        <label
          for="server"
        >
          Server Command
        </label>
        <i
          class="bar"
        />
      </div>
      <button
        aria-disabled="false"
        class="sc-bdVaJa dmOrLZ"
        type="submit"
      >
        Submit
      </button>
    </form>
  </div>
</div>
`;
